### DISCUSSION TITLE
Zlib dependency issue

### DISCUSSION THREAD

**Discussion by jrmcmah2** [python-pillow/Pillow#7265] [Q&A] [answered]
I am using the following command to install Pillow on my local windows 10 machine using Python 3.9: 

python3 -m pip install --upgrade Pillow --no-binary :all:

When I do this I am receiving an error that the zlib files cannot be found:

![image](https://github.com/python-pillow/Pillow/assets/40073416/c2b75e84-388e-4f59-93fc-16b8a26b5f26)

I do have Zlib installed at: C:\Program Files (x86)\Zlib

![image](https://github.com/python-pillow/Pillow/assets/40073416/04aa25f3-e967-4600-885b-dd2c88f605e4)

I have been looking at the installation page for guidance but cannot find where the wheel for Pillow expects the Zlib files to be. Can someone provide guidance here?

**Comment by radarhere** [comment]
Using a wheel shouldn't mean that you need to have zlib installed. It should be either that you are using a wheel, or that you need zlib installed, not both.

Could you provide the full output of the pip command?

**Reply by jrmcmah2** [reply]
So I first ran "python -m pip install --upgrade pip" and that ran just fine. Then I ran "python -m pip install --upgrade Pillow --no-binary :all:"

The output of that second command was as follows:
Collecting Pillow
  Using cached Pillow-10.0.0.tar.gz (50.5 MB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Building wheels for collected packages: Pillow
  Building wheel for Pillow (pyproject.toml): started
  Building wheel for Pillow (pyproject.toml): finished with status 'error'
Failed to build Pillow

  error: subprocess-exited-with-error
  
  Building wheel for Pillow (pyproject.toml) did not run successfully.
  exit code: 1
  
  [201 lines of output]
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win32-cpython-39
  creating build\lib.win32-cpython-39\PIL
  copying src\PIL\BdfFontFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\BlpImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\BmpImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\BufrStubImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ContainerIO.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\CurImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\DcxImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\DdsImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\EpsImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ExifTags.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\features.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\FitsImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\FliImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\FontFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\FpxImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\FtexImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\GbrImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\GdImageFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\GifImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\GimpGradientFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\GimpPaletteFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\GribStubImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\Hdf5StubImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\IcnsImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\IcoImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\Image.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageChops.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageCms.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageColor.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageDraw.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageDraw2.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageEnhance.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageFilter.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageFont.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageGrab.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageMath.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageMode.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageMorph.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageOps.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImagePalette.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImagePath.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageQt.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageSequence.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageShow.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageStat.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageTk.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageTransform.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImageWin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\ImtImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\IptcImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\Jpeg2KImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\JpegImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\JpegPresets.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\McIdasImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\MicImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\MpegImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\MpoImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\MspImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PaletteFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PalmImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PcdImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PcfFontFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PcxImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PdfImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PdfParser.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PixarImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PngImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PpmImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PsdImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PSDraw.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\PyAccess.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\QoiImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\SgiImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\SpiderImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\SunImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\TarIO.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\TgaImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\TiffImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\TiffTags.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\WalImageFile.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\WebPImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\WmfImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\XbmImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\XpmImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\XVThumbImagePlugin.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\_binary.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\_deprecate.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\_tkinter_finder.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\_util.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\_version.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\__init__.py -> build\lib.win32-cpython-39\PIL
  copying src\PIL\__main__.py -> build\lib.win32-cpython-39\PIL
  running egg_info
  writing src\Pillow.egg-info\PKG-INFO
  writing dependency_links to src\Pillow.egg-info\dependency_links.txt
  writing requirements to src\Pillow.egg-info\requires.txt
  writing top-level names to src\Pillow.egg-info\top_level.txt
  reading manifest file 'src\Pillow.egg-info\SOURCES.txt'
  reading manifest template 'MANIFEST.in'
  warning: no files found matching '*.c'
  warning: no files found matching '*.h'
  warning: no files found matching '*.sh'
  warning: no files found matching '*.txt'
  warning: no previously-included files found matching '.appveyor.yml'
  warning: no previously-included files found matching '.clang-format'
  warning: no previously-included files found matching '.coveragerc'
  warning: no previously-included files found matching '.editorconfig'
  warning: no previously-included files found matching '.readthedocs.yml'
  warning: no previously-included files found matching 'codecov.yml'
  warning: no previously-included files found matching 'renovate.json'
  warning: no previously-included files matching '.git*' found anywhere in distribution
  warning: no previously-included files matching '*.pyc' found anywhere in distribution
  warning: no previously-included files matching '*.so' found anywhere in distribution
  no previously-included directories found matching '.ci'
  adding license file 'LICENSE'
  writing manifest file 'src\Pillow.egg-info\SOURCES.txt'
  warning: build_py: byte-compiling is disabled, skipping.
  
  running build_ext
  
  
  The headers or library files could not be found for zlib,
  a required dependency when compiling Pillow from source.
  
  Please see the install instructions at:
     https://pillow.readthedocs.io/en/latest/installation.html
  
  Traceback (most recent call last):
    File "<string>", line 988, in <module>
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\__init__.py", line 107, in setup
      return distutils.core.setup(**attrs)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 185, in setup
      return run_commands(dist)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\core.py", line 201, in run_commands
      dist.run_commands()
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 969, in run_commands
      self.run_command(cmd)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\dist.py", line 1234, in run_command
      super().run_command(command)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\wheel\bdist_wheel.py", line 343, in run
      self.run_command("build")
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 318, in run_command
      self.distribution.run_command(command)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\dist.py", line 1234, in run_command
      super().run_command(command)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\command\build.py", line 131, in run
      self.run_command(cmd_name)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\cmd.py", line 318, in run_command
      self.distribution.run_command(command)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\dist.py", line 1234, in run_command
      super().run_command(command)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\command\build_ext.py", line 84, in run
      _build_ext.run(self)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\_distutils\command\build_ext.py", line 345, in run
      self.build_extensions()
    File "<string>", line 811, in build_extensions
  __main__.RequiredDependencyException: zlib
  
  During handling of the above exception, another exception occurred:
  
  Traceback (most recent call last):
    File "C:\Users\jimmy\Desktop\Projects\Runescape\venv\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
      main()
    File "C:\Users\jimmy\Desktop\Projects\Runescape\venv\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "C:\Users\jimmy\Desktop\Projects\Runescape\venv\lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 251, in build_wheel
      return _build_backend().build_wheel(wheel_directory, config_settings,
    File "C:\Users\jimmy\AppData\Local\Temp\pip-install-1q6cwqnm\pillow_be6620f1fc8b48dea5bd97c092b1c0b2\_custom_build\backend.py", line 53, in build_wheel
      return super().build_wheel(wheel_directory, config_settings, metadata_directory)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\build_meta.py", line 416, in build_wheel
      return self._build_with_temp_dir(['bdist_wheel'], '.whl',
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\build_meta.py", line 401, in _build_with_temp_dir
      self.run_setup()
    File "C:\Users\jimmy\AppData\Local\Temp\pip-install-1q6cwqnm\pillow_be6620f1fc8b48dea5bd97c092b1c0b2\_custom_build\backend.py", line 47, in run_setup
      return super().run_setup(setup_script)
    File "C:\Users\jimmy\AppData\Local\Temp\pip-build-env-wwiwcrx4\overlay\Lib\site-packages\setuptools\build_meta.py", line 338, in run_setup
      exec(code, locals())
    File "<string>", line 1005, in <module>
  __main__.RequiredDependencyException:
  
  The headers or library files could not be found for zlib,
  a required dependency when compiling Pillow from source.
  
  Please see the install instructions at:
     https://pillow.readthedocs.io/en/latest/installation.html
  
  
  [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for Pillow
ERROR: Could not build wheels for Pillow, which is required to install pyproject.toml-based projects

**Reply by jrmcmah2** [reply]
Is there a different command that I should run here to be able to disregard the zlib dependencies? I try to install it through pycharm as well and the output still ends with issues regarding the zlib dependency.

**Reply by aclark4life** [reply]
You cannot disregard zlib dependencies when compiling, and compiling is attempted when no wheels are available for your architecture …

**Reply by jrmcmah2** [reply]
Got it, @aclark4life. Do I need to install something or run some other pip command to obtain the proper wheels? In the installation docs it says I need both Zlib and libjpeg. I have installed (Or at least thought I have installed) both of them.

**Reply by aclark4life** [reply]
Installation via compilation requires zlib, jpeg, etc. When compilation succeeds, the results are used to create a wheel. Ideally there's already a wheel for your architecture and you don't need to compile.

**Comment by homm** [comment]
`Program Files (x86)` is a 32-bit, right? What is your `Python 3.9` architecture?

**Reply by jrmcmah2** [reply]
Sorry can you clarify what you mean by Python3.9 Architecture?

**Reply by aclark4life** [reply]
Where is your Python 3.9 installed?

**Reply by jrmcmah2** [reply]
C:\\Users\\UserName\\AppData\\Local\\Programs\\Python\\Python39-32

**Comment by nulano** [comment, marked as answer]
AFAIK dependencies aren't detected automatically on Windows, you have to add them to the INCLIB envvar before attempting to compile Pillow from source. A script that compiles all dependencies and sets the required envvars is provided in the https://github.com/python-pillow/Pillow/tree/main/winbuild directory.

**Reply by C0rn3j** [reply]
I was installing via pip, so had to export the include lib before running pip. LDFLAGS was not necessary but leaving it here just in case.

```bash
export CFLAGS="-I/mingw64/include"
#export LDFLAGS="-L/mingw64/lib"
```
